Chore: Convert ArrayField to functional components (#4808)
rjsf-team/react-jsonschema-form: packages/docs/docs/migration-guides/v6.x upgrade guide.md

### Templates BREAKING CHANGES

> #### ArrayFieldTemplateItemType

Author: Also, rename this to `ArrayFieldItemTemplateProps`

```suggestion
#### ArrayFieldTemplateItemType POTENTIAL BREAKING CHANGE
```

- `hasMoveDown`: A boolean value stating whether the array item can be moved down.
- `hasMoveUp`: A boolean value stating whether the array item can be moved up.
- `hasRemove`: A boolean value stating whether the array item can be removed.
- ~~`onAddIndexClick: (index) => (event?) => void`: Returns a function that adds a new item at `index`.~~
- ~~`onCopyIndexClick: (index: number) => (event?: any) => void;`: Returns a function that copies the item at `index` into the position at `index + 1`.~~
- ~~`onDropIndexClick: (index) => (event?) => void`: Returns a function that removes the item at `index`.~~
- ~~`onReorderClick: (index, newIndex) => (event?) => void`: Returns a function that swaps the items at `index` with `newIndex`.~~

> In addition to this refactor, the conversion of the `ArrayField` template to a functional component caused the replacement of the 4 callback-generator functions with 5 memoizable functions.

Author: ```suggestion
In addition to this refactor, the conversion of the `ArrayField` template to functional components caused the replacement of the 4 callback-generator functions with 5 memoizable functions.
```

                onClick={element.onReorderClick(element.index, element.index + 1)}
              >
                Down
              </button>
            )}
            {(element.hasMoveUp || element.hasMoveDown) && (
              <button className='array-item-move-up' onClick={element.onReorderClick(element.index, element.index - 1)}>
                Up
              </button>
            )}
            {element.hasCopy && (
              <button className='array-item-copy' onClick={element.onCopyIndexClick(element.index)}>
                Copy
              </button>
            )}
            {element.hasRemove && (
              <button className='array-item-remove' onClick={element.onDropIndexClick(element.index)}>
                Remove
              </button>
            )}
            <hr />
          </div>
        ))}
      {canAdd && (
        <div className='array-item-add'>
          <button onClick={onAddClick} type='button'>
            Add New
          </button>
        </div>
      )}
    </div>
  );
}
```

```tsx
// v6 ArrayFieldTemplate and ArrayFieldItemTemplate (including other v6 changes mentioned in other sections)

> import { ArrayFieldTemplateProps } from '@rjsf/utils';

Author: ```suggestion
import { ArrayFieldItemTemplateType, ArrayFieldTemplateProps } from '@rjsf/utils';
```

> function CustomArrayFieldTemplate(props: ArrayFieldTemplateProps) {

Author: ```suggestion
function CustomArrayFieldItemTemplate(props: ArrayFieldItemTemplateType) {
```

The following new types were added to `@rjsf/utils`:

> - `ArrayFieldItemTemplateProps`: The properties of each element in the ArrayFieldTemplateProps.items array. NOTE: `ArrayFieldTemplateItemType` is an alias to this type

Reviewer: ```suggestion
- `ArrayFieldItemTemplateProps`: The properties of each element in the ArrayFieldTemplateProps.items array.
```
